### PR TITLE
Mariadb JDBC plugin

### PR DESCRIPTION
**Pull request by efekaptan**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
#3907 

___
### New feature or improvement
- A JDBC driver plugin implemented for Mariadb. As Mariadb is based on the Mysql project, there are similarities between two drivers. 
- Mariadb uses the Mysql url format so I re-used the `MysqlURLParser` with a small modification.
- PreparedStatement parameter string building logic is exactly the same between Mysql, PostgreSql and Mariadb. I created a parameter builder class to re-use between these plugins. That's why I touched a little bit to Postgresql and Mysql plugins and I checked the unit tests and scenarios to be sure to not break anything.
- The latest minor versions of the the driver(2.x) are included in the test scenario.

image

### CODE REVIEW

File: oap-server/server-bootstrap/src/main/resources/component-libraries.yml [hunk 1]
```diff
@@ -287,6 +287,9 @@ Undertow:
 Finagle:
   id: 85
   languages: Java,Scala
+Mariadb:
```
Comment (wu-sheng) [inline oap-server/server-bootstrap/src/main/resources/component-libraries.yml:290]: I think you need a logo PR at the rocketbot UI repo.

One question, does `Mariadb` or `Mariadb-JDBC` make more sense? if `Mariadb-JDBC`, then should add mapping to `Component-Server-Mappings:`

File: apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/ComponentsDefine.java [hunk 2]
```diff
@@ -156,4 +156,6 @@
     public static final OfficialComponent UNDERTOW = new OfficialComponent(84, "Undertow");
 
     public static final OfficialComponent FINAGLE = new OfficialComponent(85, "Finagle");
+
+    public static final OfficialComponent MARIADB = new OfficialComponent(86, "Mariadb");
```
Comment (wu-sheng) [inline apm-protocol/apm-network/src/main/java/org/apache/skywalking/apm/network/trace/component/ComponentsDefine.java:160]: I think in Java, this should be `Mariadb-jdbc` or something indicating we instrumenting the client driver only.

**Comment by arugal** [comment]
test case good to me

File: oap-server/server-bootstrap/src/main/resources/component-libraries.yml (hunk 1 above)
Comment (efekaptan) [reply oap-server/server-bootstrap/src/main/resources/component-libraries.yml:290]: Ok, I agree. Other JDBC drivers are also using similar naming convention, will refactor it.

**Comment by efekaptan** [comment]
Hi, I couldn't succeed to receive the `TopNRecords` from the server. I couldn't receive for Mysql driver either. During my investigation I noticed that below line doesn't return any value:

https://github.com/apache/skywalking/blob/master/oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/worker/PersistenceWorker.java#L73

TopNWorker can't save any record. Am i missing some point? Can I get your feedback about this situation?

**Comment by wu-sheng** [comment]
Submit #4741 to fix the thing you reported, thanks.

**Comment by efekaptan** [comment]
@wu-sheng I tried again after your fix but unfortunately `TOP_N_DATABASE_STATEMENT` table remains empty and can't get any topN result from the server. I have also decreased the `topNReportPeriod` parameter to 1 minute interval.

I will compare with the demo application, would like to see if the problem is related to my setup or not.

**Comment by arugal** [comment]
> @wu-sheng I tried again after your fix but unfortunately `TOP_N_DATABASE_STATEMENT` table remains empty and can't get any topN result from the server. I have also decreased the `topNReportPeriod` parameter to 1 minute interval.
> 
> I will compare with the demo application, would like to see if the problem is related to my setup or not.

@efekaptan I have updated the master code for you, please pull the code and try again.

**Comment by wu-sheng** [comment]
@efekaptan The easiest way, use the `AgentDataMock` in the source codes, it could generate the slow SQL. Then you could see it on the UI. This is how I tested in the morning. :)

**Comment by wu-sheng** [comment]
BTW, this is how the slow SQL generated, https://github.com/apache/skywalking/blob/44cae97c259c06f95d1fc4b78ee41268670b6542/oap-server/server-receiver-plugin/skywalking-trace-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/trace/provider/parser/listener/MultiScopesAnalysisListener.java#L188-L222.

**Comment by efekaptan** [comment]
Thanks for the information, let me try to debug and understand the reason behind it.

**Comment by efekaptan** [comment]
After decreasing `slowDBAccessThreshold`, I was able to receive slow sqls from the server, thanks for the hint. 

- I refactored component name as `mariadb-jdbc`.
- I am attaching a sample database dashboard screenshot.

Screenshot 2020-05-06 at 19 49 24

File: docs/en/setup/service-agent/java-agent/README.md [hunk 3]
```diff
@@ -125,6 +125,8 @@ property key | Description | Default |
 `plugin.http.http_params_length_threshold`| When `COLLECT_HTTP_PARAMS` is enabled, how many characters to keep and send to the OAP backend, use negative values to keep and send the complete parameters, NB. this config item is added for the sake of performance.  | `1024` |
 `correlation.element_max_number`|Max element count of the correlation context.|`3`|
 `correlation.value_max_length`|Max value length of correlation context element.|`128`|
+`plugin.mariadb.trace_sql_parameters`|If set to true, the parameters of the sql (typically `java.sql.PreparedStatement`) would be collected.|`false`|
+`plugin.mariadb.sql_parameters_max_length`|If set to positive number, the `db.sql.parameters` would be truncated to this length, otherwise it would be completely saved, which may cause performance problem.|`512`|
```
Comment (wu-sheng) [inline docs/en/setup/service-agent/java-agent/README.md:129]: Please move these two just below the existing database config. Let's keep the database config in a same place

File: test/plugin/scenarios/mariadb-scenario/config/expectedData.yaml [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+segmentItems:
+- serviceName: mariadb-scenario
+  segmentSize: ge 2
+  segments:
+  - segmentId: not null
+    spans:
+    - operationName: /mariadb-scenario/case/healthCheck
+      operationId: 0
```
Comment (wu-sheng) [inline test/plugin/scenarios/mariadb-scenario/config/expectedData.yaml:23]: OperationId isn't required, this is a legacy thing, please remove all.

**Comment by wu-sheng** [comment]
The logo is for topology, you should see a question mark to represent the database node currently, yes?

File: docs/en/setup/service-agent/java-agent/README.md (hunk 3 above)
Comment (efekaptan) [reply docs/en/setup/service-agent/java-agent/README.md]: Done.

File: test/plugin/scenarios/mariadb-scenario/config/expectedData.yaml (hunk 4 above)
Comment (efekaptan) [reply test/plugin/scenarios/mariadb-scenario/config/expectedData.yaml]: Done

**Pull request merged**